internal/factsengine/gatherers/verifypassword.go

```diff
@@ -0,0 +1,98 @@
+package gatherers
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	crypt "github.com/tredoe/osutil/user/crypt"
+	sha512crypt "github.com/tredoe/osutil/user/crypt/sha512_crypt"
+	"github.com/trento-project/agent/pkg/factsengine/entities"
+	"github.com/trento-project/agent/pkg/utils"
+)
+
+const (
+	VerifyPasswordGathererName = "verify_password"
+)
+
+// nolint:gochecknoglobals
+var (
+	checkableUsernames = []string{"hacluster"}
+	unsafePasswords    = []string{"linux"}
+)
+
+// nolint:gochecknoglobals
+var (
+	VerifyPasswordInvalidUsername = entities.FactGatheringError{
+		Type:    "verify-password-invalid-username",
+		Message: "requested user is not whitelisted for password check",
+	}
+)
+
+type VerifyPasswordGatherer struct {
+	executor utils.CommandExecutor
+}
+
+func NewDefaultPasswordGatherer() *VerifyPasswordGatherer {
+	return NewVerifyPasswordGatherer(utils.Executor{})
+}
+
+func NewVerifyPasswordGatherer(executor utils.CommandExecutor) *VerifyPasswordGatherer {
+	return &VerifyPasswordGatherer{
+		executor,
+	}
+}
+
+/*
+This gatherer expects only the username for which the password will be verified
+*/
+func (g *VerifyPasswordGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	facts := []entities.Fact{}
+	log.Infof("Starting password verifying facts gathering process")
+
+	for _, factReq := range factsRequests {
+		if !utils.StringInSlice(factReq.Argument, checkableUsernames) {
+			gatheringError := VerifyPasswordInvalidUsername.Wrap(factReq.Argument)
+			fact := entities.NewFactGatheredWithError(factReq, gatheringError)
+			facts = append(facts, fact)
+			continue
+		}
+		username := factReq.Argument
+
+		salt, hash, err := g.getSalt(username)
+		if err != nil {
+			log.Error(err)
+		}
+		log.Debugf("Obtained salt using user %s: %s", username, salt)
+
+		crypter := sha512crypt.New()
+		isPasswordWeak := false
+		for _, password := range unsafePasswords {
+			passwordBytes := []byte(password)
+			match := crypter.Verify(hash, passwordBytes)
+			if !errors.Is(match, crypt.ErrKeyMismatch) {
```

Comment:
maybe an edge case but `match := crypter.Verify(hash, passwordBytes)` could return a different error than `crypt.ErrKeyMismatch`, without knowing whether the hashes are matching or not ([see here](https://github.com/tredoe/osutil/blob/master/v2/userutil/crypt/sha512_crypt/sha512_crypt.go#L221)) resulting in the detection of a weak password, which would not be necessarily true.
An option could be to handle this case explicitly and return a `FactGatheringError`. Makes sense?